### ISSUE TITLE
Adding undo and redo buttons for the in-game level editor.

### ISSUE THREAD

**Issue by IndianaTux** [SuperTux/supertux#465]

**Comment by maxteufel** [comment]
Backend suggestions:
1. The level file is created. Addtionally, a changelog file for the given level is created.
2. User changes stuff. Change is done to the level file, and a machine-readable description of what changed is written to the changelog file (that could be a diff, or an s-expression defining the change)
3. *do
3.1. Undo: we reverse the last action
3.2. Redo: we unreverse the last action
4. Repeat 3. as many times as you want

**Comment by J-128** [comment]
Yes, I find myself wanting to hit Ctrl-Z a lot, so Undo/Redo would be nice.

**Comment by Karkus476** [comment]
@maxteufel Do we need to preserve these after the program closes?

**Comment by maxteufel** [comment]
@Karkus476 It wouldn't have to be preserved. We should even add a limit during runtime to prevent memory overuse.

**Comment by Karkus476** [comment]
@maxteufel Why save it in a file then?

**Comment by maxteufel** [comment]
"file". Everything on UNIX is a file :smile:. Anyway, it could be saved in RAM, too. But note that modern applications often also save edit history in case of crashes.

**Comment by Karkus476** [comment]
@maxteufel Okay, that's a good idea. By the way I'm not trying to be annoying, just considering the way flexlay achieves this. It saves objects with redo and undo methods in a list.

**Comment by maxteufel** [comment]
@Karkus476 the more interesting thing to know would be how flexlay serializes the undo/redo data.

**Comment by tobbi** [comment]
I'm actually thinking of having a stack with custom "ChangeItem" objects that describe the change.

ChangeObject:
{
type: changeobject_t,
state_before: ...
state_after: ...
}

Or maybe they have on_undo and on_redo functions.
